core/src/main/scala/org/apache/spark/deploy/security/README.md

```diff
@@ -0,0 +1,238 @@
+# Delegation Token Handling In Spark
+
+This document aims to explain and demystify delegation tokens as they are used by Spark, since
+this topic is generally a huge source of confusion.
+
+
+## What are delegation tokens?
+
+Delegation tokens (DTs from now on) are authentication tokens used by some services to replace
+Kerberos service tokens. Many services in the Hadoop ecosystem have support for DTs, since they
+have two very desirable advantages over Kerberos tokens:
+
+* No need to distribute Kerberos credentials
+
+In a distributed application, distributing Kerberos credentials is tricky. Not all users have
+keytabs, and when they do, it's generally frowned upon to distribute them over the network as
+part of application data.
+
+DTs allow for a single place (e.g. the Spark driver) to require Kerberos credentials. That entity
```

Comment:
this isn't true, delegation tokens don't allow for single place.  you could just as well have a keytab in the single place that distributes the tgts. And really its not a single place since it distributes them to executors.  We might rephrase this as an implementation detail that they go to the driver and are distributed

Reply:
>  delegation tokens don't allow for single place

Hmm, yes they do. Yes, you could do those other things, but then you wouldn't need DTs. So DTs are one way to allow a single place to handle kerberos credentials.

>  delegation tokens don't allow for single place

Not sure I follow. Only a single place has *kerberos* credentials. DTs are distributed to executors, but they are not *kerberos* credentials.

Reply:
Oh I think I see what you are trying to say here.  I think you need to elaborate more on this.  I think you are trying to say that if you are sending kerberos credentials like a keytab that allows us to do all DT handling in a single place - the driver.

Reply:
No, what I'm trying to say is that when you use DTs, there's a single place that needs to handle kerberos credentials.

It doesn't matter if you're using keytabs or kinit. It's still a single place. And that single place is responsible for distributing the DTs to the other entities that needs to authenticate against services.

It also doesn't need to be the Spark driver. As I mention in other places in the document, the DTs can be created by an external entity (like Oozie does).

Reply:
ok, sorry I understand now, I somehow really misread this.